exercises/e001/results/mikjarz/e001.py

```diff
@@ -0,0 +1,33 @@
+import random
+import string
+import base64
+import argparse
+
+parser = argparse.ArgumentParser(description="Random string of base32 characters")
+parser.add_argument('-a', '--amount', type=int, help='How many characters to print')
+parser.add_argument('-u', '--upper', action='store_true', help='Printing all upper characters')
+parser.add_argument('-l', '--lower', action='store_true', help='Printing all lower characters')
+args = parser.parse_args()
+
+
+def random_string(n):
+    letters = string.ascii_letters + string.digits
+    zz = (''.join(random.choice(letters) for i in range(n)))
+    encoded = zz.encode("utf-8")
+    b32coded = base64.b32encode(encoded)
```

Comment:
from python documentation:

```
base64.b32encode(s)
  Encode the bytes-like object s using Base32 and return the encoded bytes.
```

If `b32encode` takes bytes-like object then you should supply byte like object with full spectrum of possibilities. A bytes is 8-bits, so 2^8 = 256 possibilities.

However you call `base64.b32encode(encoded)` with `encoded` derived from `letters = string.ascii_letters + string.digits` which is `62` possibilities `len(string.ascii_letters + string.digits)`

Which means your randomness is with lower entropy space (256 versus 62). You don't want that. Your password is weaker.

You want to generate random from full space of `s` of `base64.b32encode(s)` which means you should generate random via `random.choice(range(0,256))` or via `random.randint(0, 255)` (basically you generate random `byte`, because `b32encode` takes bytes-like object).

```
     1  import random
     2  import base64
     3  
     4  def random_string(n):
     5      r = random.randint(0, 255)
     6      a = [random.randint(0, 255) for _ in range(n)]
     7      b = bytes(a)
     8      return base64.b32encode(b)
     9  
    10  if __name__ == '__main__':
    11      n = 32
    12      s = random_string(n)[:n]
    13      print(s.decode('ascii'))
```

Lets analyze above. Starting from line 5 of `random_string()` function. Line 5 generate random intiger between 0 and 255 (8-bits of a byte). Then we generate array of random 8-bits ints of length `n` on line 6. On line 7 we convert this array of ints into byte array. Then we can encode this byte array with `b32encode` which returns also byte array but base32 encoded.

Because base32 (32 possibilities) is smaller space than what `random.randint(0, 255)` is (256 possibilities) we can safely truncate byte array to our final length via `[:n]` slice of array notation, on line 12.

As all the functions so far worked on byte array and `print()` on line 13 would add byte array notation to `b''` to our string which we don't want, we `decode` the array into ASCII string. I used ASCII because I know `base32` fits into ASCII.

Happy to discuss over the phone if needed. Don't hesitate to reach out.